forest/cli/src/cli/snapshot_cmd.rs

```diff
@@ -323,6 +349,130 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
+            .expect("Opening RocksDB must succeed");
+
+        let chain_store = Arc::new(ChainStore::new(db).await);
+
+        let genesis = read_genesis_header(
+            config.client.genesis_file.as_ref(),
+            config.chain.genesis_bytes(),
+            &chain_store,
+        )
+        .await
+        .unwrap();
+        let genesis_header = &genesis.blocks()[0];
+        chain_store.set_genesis(genesis_header).unwrap();
+
+        let cids = {
+            let file = tokio::fs::File::open(&snapshot).await?;
+            let reader = FetchProgress::fetch_from_file(file).await?;
+            load_car(chain_store.blockstore(), reader.compat()).await?
+        };
+
+        let ts = chain_store.tipset_from_keys(&TipsetKeys::new(cids)).await?;
+
+        validate_links_and_genesis_traversal(
+            &chain_store,
+            ts,
+            chain_store.blockstore(),
+            *recent_stateroots,
+            &genesis,
+        )
+        .await?;
+    }
+
+    Ok(())
+}
+
+async fn validate_links_and_genesis_traversal<DB>(
+    chain_store: &ChainStore<DB>,
+    ts: Arc<Tipset>,
+    db: &DB,
+    recent_stateroots: ChainEpoch,
+    genesis_tipset: &Tipset,
+) -> anyhow::Result<()>
+where
+    DB: fvm_ipld_blockstore::Blockstore + Store + Send + Sync,
+{
+    let mut seen = std::collections::HashSet::<Cid>::new();
+    let upto = ts.epoch() - recent_stateroots;
+
+    println!(
+        "validating cids from snapshot in db: {} <- {} ...",
+        upto,
+        ts.epoch()
+    );
+
+    let mut tsk = ts.parents().clone();
+    // Security: Recursive snapshots are difficult to create but not impossible. This
+    // limits the amount of recursion we do.
+    let mut prev_epoch = ts.epoch();
+
+    loop {
+        if prev_epoch <= 0 {
+            bail!("could not reach genesis from snapshot");
+        }
+
+        let tipset = chain_store.tipset_from_keys(&tsk).await?;
+        if tipset.epoch() >= prev_epoch {
+            bail!(
+                "tipset iteration didn't make progress backwards current epoch: {} previous_epoch: {}, possible recursion in snapshot.", tipset.epoch(), prev_epoch
+            );
+        }
+        let height = tipset.epoch();
+        if height == 0 {
+            if tipset.as_ref() != genesis_tipset {
+                bail!("could not reach genesis from snapshot, genesis tipset in snapshot doesn't match the reference");
+            }
+
+            break;
+        }
+        if height > upto {
+            let mut assert_cid_exists = |cid: Cid| async move {
+                let data = db.get(&cid)?;
+                data.ok_or_else(|| {
+                    anyhow::anyhow!(
+                        "could not find data for cids in tipset at height: {}",
+                        height
+                    )
+                })
+            };
+
+            for h in tipset.blocks() {
+                recurse_links(&mut seen, *h.state_root(), &mut assert_cid_exists).await?;
```

Comment:
You also need to check the links in `h.messages()`, btw.

Reply:
added